### ISSUE TITLE
`swayidle ... resume 'swaymsg "output * dpms on"'` drops repeated keys

### ISSUE THREAD

**Issue by camoz** [swaywm/sway#7112]
**Sway Version:** 1.7

**Debug Log:** https://gist.github.com/camoz/ce7bc5e8385418cd2eaf47a89f2153ee

**Minimal reproducing config:** Default config with this line added:

`exec swayidle -w timeout 1 'true' resume 'swaymsg "output * dpms on"'`

**Description:**

With the line above part of the sway config, when I keep inputs idle for 1+ seconds and hold down a key (e.g. in a terminal emulator), only one character gets printed. When I hold down a key before the computer has idled for 1+ seconds, they key is repeated as usual. This happens in foot, termite (VTE-based terminal emulator), and Qt apps (like wireshark).

Here is the output of `WAYLAND_DEBUG=1 foot` losing the repeated keys: https://gist.github.com/camoz/1848fe4fd08b2c86c7feafb817f20a03

The foot log and sway debug log are from the same sway session.

How I obtained the logs:

- Add mapping `bindsym $mod+Return exec WAYLAND_DEBUG=1 $term > /tmp/foot.log 2>&1` to minimal sway config posted above.
- Start sway
- Hit Mod+Return
- Wait for 1+ seconds
- Hold down `j` for 2+ seconds
- Release `j`
- Press Mod+Shift+E to exit sway
- Click `exit sway`

**Comment by MayeulC** [comment]
I have also experienced that bug recently, intermittently. Sway 1.7, Arch Linux.
I try to avoid swayidle these days because I'm encountering multiple similar bugs: keys not repeating, existing windows unable to process input on resume, unable to change VTs, `dpms on` not activated on resume, and counting...

**Comment by camoz** [comment]
I just noticed that I also get duplicate keys when just briefly pressing a key:

1. open terminal, run `swayidle -w -d timeout 4 'swaymsg "output * power off"' resume 'swaymsg "output * power on"'`
2. wait 4+ seconds
3. press a key (that echoes in the terminal, e.g. a letter)
4. observe how the screen turns back on, and the letter is inserted *multiple times*

This makes it annoying to unlock my screen, because when the screen is locked (via swaylock) and the screen is powered off, then I cannot just type in my password; I first have to power on the screen by pressing e.g. shift, and then enter the password. Otherwise, the first character of my password is inserted multiple times, which obviously will not pass/unlock.

**Comment by kennylevinsen** [comment]
This is unrelated to swayidle. The cause is that you press the key, and afterwards monitors start to enable and modeset. This process takes a while, during which input is not processed. Only when the modesets finish is the key release sent.

Clients that received the initial key-press will see this as the key being held for a prolonged period, and repeat the key as they are configured to do. They are unaware that the key release is merely delayed because sway is busy, not because the key was not yet released.

**Comment by camoz** [comment]
I see, thanks for the clarification. @kennylevinsen Do you have a suggestion for a better title?

**Comment by pielgrzym** [comment]
@kennylevinsen thanks for this explanation! So power setting for output is a blocking operation in sway - perhaps we could just use other utility to shut  down monitor via dpms? AFAIK `wlr-randr` can do this - will it work with sway?

**Comment by kennylevinsen** [comment]
No, sway is the DRM master and is the only process that can control the outputs. Tools like swaymsg, wlr-randr, kanshi etc. just tell sway what they would like done.

**Comment by pielgrzym** [comment]
I discovered why the DPMS is not always working. When a xwayland window is active (or sometimes even on given display) dpms won't start. When I move to workspace without any XWayland apps (or an empty one for that matter) dpms locking always works.

**Comment by ewtoombs** [comment]
This is affecting screen unlocking.
If a password is typed in while the screen is in the dpms off state, the first letter is repeated once or twice and the login fails.
As a workaround, for years, I have had to wake the screen first with, for example, the shift key before typing in my password.

Here is my minimal steps for reproducing:
0. Set the keyboard_delay *really* low, but still usable. 100 milliseconds is fine.
1. Send a `dpms on` command first to make it easier to turn the screen back on again.
2. Send the `dpms off` command.
3. Your screen is off now, so you can't see. But your terminal is still working fine. Rerun that `dpms on` command from earlier, by carefully pressing *up, up, enter*.

Here's what it looks like:
```
eric@robert ~> swaymsg 'input type:keyboard repeat_delay 100'
eric@robert ~> swaymsg 'output * dpms on'
eric@robert ~> swaymsg 'output * dpms off'










eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> swaymsg 'output * dpms on'









eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
eric@robert ~> 
```
Interesting! As you can see, even the dpms off command is causing keyboard input processing to hang.
The enter key gets repeated about nine times instead of just once.

I have seen similar results with `repeat_delay` as high as 250 milliseconds, which is well within the range that actual humans would use. (I use 250 ms myself.)

**Comment by kennylevinsen** [comment]
As established above, this behavior is well known. The display driver can end up blocking for an arbitrary duration when told to change display configuration, which in turn freezes sway for that period. If this happens while a key is pressed down, the key lift is delayed.

**Comment by emersion** [comment]
The fix would be to switch to non-blocking modesets, but there are a number of kernel-side issues with these.

**Comment by ewtoombs** [comment]
Another solution could be to not count the time in the blocking call towards the repeat_delay.
